edb/edgeql/declarative.py

```diff
@@ -1200,6 +1202,39 @@ def _resolve_type_expr(
         )
 
 
+TRACER_TO_REAL_TYPE_MAP = {
+    qltracer.ObjectType: s_objtypes.ObjectType,
+    qltracer.ScalarType: s_scalars.ScalarType,
+    qltracer.Constraint: s_constr.Constraint,
+    qltracer.Annotation: s_anno.Annotation,
+    qltracer.Property: s_props.Property,
+    qltracer.Link: s_links.Link,
+}
+
+
+def _get_local_obj(
+    refname: s_name.QualName,
+    tracer_type: Type[qltracer.NamedObject],
+    sourcectx: Optional[parsing.ParserContext],
+    *,
+    ctx: LayoutTraceContext,
+) -> Optional[qltracer.NamedObject]:
+
+    obj = ctx.objects.get(refname)
+
+    if obj is not None and not isinstance(obj, tracer_type):
+        obj_type = TRACER_TO_REAL_TYPE_MAP[type(obj)]
+        real_type = TRACER_TO_REAL_TYPE_MAP[tracer_type]
+        raise errors.InvalidReferenceError(
+            f'{str(refname)!r} exists, but is a '
+            f'{obj_type.get_schema_class_displayname()!r}, '
```

Comment:
Hm, perhaps we could lose the quotes here.  `'test::X' exists, but is a scalar type, not a object type` reads better to me (minus a more obvious article error).

Reply:
Sure

Reply:
I fixed the articles too